### ISSUE TITLE
Run ElastAlert rule only once

### ISSUE THREAD

**Issue by adminnz** [Yelp/elastalert#1775]
I'm looking to run multiple ElastAlert instances each as Azure Functions with different rules triggered by external independent timers.
Is there a way to have ElastAlert run a specific action file just once and then exit? i.e. I dont want to utilize the run_every parameter and utilize my external timers to handle when ElastAlert performs the query against ElasticSearch.

**Comment by Qmando** [comment]
If you provide ``--end``, it will exit when it gets to that time. So doing something like ``elastalert --verbose --end $(date --iso-8601=seconds)`` will make elastalert run from the last end time up to the present time, then stop.

**Comment by adminnz** [comment]
I have managed to install ElastAlert (pip install elastalert) into a python Azure Function.
Is it possible to run the ElastAlerter class without passing in config & action files and instead programatically set the values that the config/action files would of been parsed into?

**Comment by Qmando** [comment]
Not really. It's technically possible of course, but very difficult.

Do you have access to making temp files?

**Comment by g-io** [comment]
The solution proposed with the `--end`-flag seems not to be working anymore. If I get it right, the problem is that with the advent of elastalert 2.0 the job execution has been delegated to apscheduler's BackgroundScheduler. So in this case, the call to `self.scheduler.start()`is returning immediately and elastalert is exiting before the jobs have been run.

This is very unfortunate as this also breaks the 'run elastalert in Lambda-function' use-case.
Not sure how to run elastalert once now. Any clue would be much appreciated.

**Comment by Qmando** [comment]
@g-io 
I haven't tested this myself but the code looks like it supports this still.
```
           next_run = datetime.datetime.utcnow() + self.run_every
            # Quit after end_time has been reached
            if self.args.end:
                endtime = ts_to_dt(self.args.end)
                if next_run.replace(tzinfo=dateutil.tz.tzutc()) > endtime:
                    exit(0)
```
Perhaps it's exiting immediately for you because the --end flag is too close to the current time. Can you try making sure that it has at least one full `run_every` between the start and end times?

**Comment by g-io** [comment]
@Qmando yes good point, that would work (although it needs to be at least `run_every` between now and end time it seems?), I've tested it yesterday already after looking into the code. However, *would* work in most cases and doesn't support the Lambda case really well.

If I get it right (haven't looked into the code but from my experience of running elastalert 1.x on Lambda) the semantics before was: Run until all jobs which have been gathered are finished, and end-time reached. This was ideal for the Lambda case, as I could let the Lambda run with `--end=<now>` every few minutes, to process the last minutes. The runtime of the Lambda then is usually short, say a few seconds, because elastalert would exit. This leverages the operational savings of running it in Lambda.

Allowing the Lambda to run longer was helping the case of a longer timeframe to cover or a peak in records to query, which I could do because elastalert would quit when the records had been processed, might that be many or few.

Now the processing is done in a different thread from the end-condition checking without any communication about the status of the jobs. Any job could be stopped while running, no matter how wisely I try to choose the runtime by adjusting `run_every` and `--end`. So either I try to choose a small timeframe (say 30sec) and hope that usually all jobs will finish in that time or I choose a long one (say 3min) and I will overpay because my Lambda is still running all the time with my elastalert sleeping. (Overpaying will always happen in this scenario because I have to idly wait until check_every has passed and end is reached, it's more a matter of how much, but well, as long as it's just some seconds per minute...)

Also I'm wondering, will elastalert jobs tolerate being stopped by the main thread repeatedly, which could happen? Will the status always be consistent so that I could safely start fresh where before a process has been interrupted? I've come across weird cases with eg
`RuntimeError: cannot schedule new futures after shutdown` playing with  `run_every` and `--end` numbers.

I might be missing something obvious, but decoupling the jobs (running in the background) from the exit condition checking seems harmful.

**Comment by briandefiant** [comment]
> The solution proposed with the `--end`-flag seems not to be working anymore. If I get it right, the problem is that with the advent of elastalert 2.0 the job execution has been delegated to apscheduler's BackgroundScheduler. So in this case, the call to `self.scheduler.start()`is returning immediately and elastalert is exiting before the jobs have been run.
> 
> This is very unfortunate as this also breaks the 'run elastalert in Lambda-function' use-case. Not sure how to run elastalert once now. Any clue would be much appreciated.

I ran into this same issue. I ended up extending the `ElastAlerter` class and overriding the `start()` method to get it working. Details are here: https://github.com/jertel/elastalert2/discussions/1169
